sharktank/tests/models/llama/kv_cache_test.py

```diff
@@ -0,0 +1,258 @@
+                    use_hf=False,
+                )
+                for n in range(self.block_count)
+            ]
+        )
+        self.direct_attn_blocks = nn.ModuleList(
+            [
+                PagedLlamaAttentionBlock(
+                    theta=self.attention_block_theta,
+                    block_index=n,
+                    cache=self.direct_kv_cache,
+                    head_count=self.head_count,
+                    head_dim=self.head_dim,
+                    head_count_kv=self.head_count_kv,
+                    rms_epsilon=self.rms_epsilon,
+                    use_hf=False,
+                )
+                for n in range(self.block_count)
+            ]
+        )
+        self.paged_cache_state = self.paged_kv_cache.allocate(page_count=128)
+        self.paged_seq_block_ids = torch.tensor(
+            [
+                [127],
+            ]
+        )
+        self.direct_cache_state = self.direct_kv_cache.allocate(bs=1)
+        self.direct_seq_block_ids = torch.tensor(
+            [
+                [0],
+            ]
+        )
+        self.embedding_batch_mask = self.attention_embedding.compute_batch_mask(
+            self.start_positions, batch_seq_len=1
+        )
+        self.model = causal_llm.BaseCausalLMModel(
+            self.attention_block_theta, context_length=self.max_seq_len
+        )
+        self.prefill_attention_mask = self.model.attention_mask(
+            self.model.input_mask(self.start_positions, self.seq_len)
+        )
+
+    def testDirectAndPagedKVCachePrefill(self):
+        torch.set_default_dtype(torch.float32)
+
+        paged_input_tensor = make_rand_torch(
+            (1, self.seq_len, self.head_count * self.head_dim), dtype=torch.float32
+        )
+        direct_input_tensor = paged_input_tensor.detach().clone()
+        # Iterate over paged attention blocks.
+        for block_idx, paged_block in enumerate(self.paged_attn_blocks):
+            paged_input_tensor = paged_block(
+                paged_input_tensor,
+                embedding=self.attention_embedding,
+                start_index=0,
+                attention_mask=self.prefill_attention_mask,
+                cache_state=self.paged_cache_state,
+                seq_block_ids=self.paged_seq_block_ids,
+            )
+        # Iterate over direct attention blocks.
+        for block_idx, direct_block in enumerate(self.direct_attn_blocks):
+            direct_input_tensor = direct_block(
+                direct_input_tensor,
+                embedding=self.attention_embedding,
+                start_index=0,
+                attention_mask=self.prefill_attention_mask,
+                cache_state=self.direct_cache_state,
+                seq_block_ids=self.direct_seq_block_ids,
+            )
+        page_table = self.paged_kv_cache.unflatten_page_table(self.paged_cache_state)
+        index_written = self.start_positions.item()
+        page_id = self.paged_seq_block_ids[0][0].item()
```

Comment:
would be useful for people less familiar with how this works to add a comment explaining the indexing you're doing here and why

Reply:
Done